core-active-state.html

```diff
@@ -0,0 +1,175 @@
+<script>
+  Polymer({
+    is: 'core-active-state',
+
+    properties: {
+
+      /**
+       * If true, the element is currently active either because the
+       * user is touching it, or the button is a toggle
+       * and is currently in the active state.
+       *
+       * @attribute active
+       * @type boolean
+       * @default false
+       */
+      active: {
+        type: Boolean,
+        notify: true,
+        readOnly: true,
+        observer: 'activeChanged',
+        value: false
+      },
+
+      /**
+       * If true, the user is currently holding down the button.
+       *
+       * @attribute pressed
+       * @type boolean
+       * @default false
+       */
+      pressed: {
+        type: Boolean,
+        value: function() {
+          this._setPressed(this.pressed || false);
+        },
+        readOnly: true,
+        observer: 'pressedChanged'
+      },
+
+      /**
+       * If true, the button toggles the active state with each tap.
+       * Otherwise, the button becomes active when the user is holding
+       * it down.
+       *
+       * @attribute toggle
+       * @type boolean
+       * @default false
+       */
+      toggle: {
+        type: Boolean,
+        value: false
+      },
+
+      /**
+       * If true, the user cannot interact with this element.
+       *
+       * @attribute disabled
+       * @type boolean
+       * @default false
+       */
+      disabled: {
+        type: Boolean,
+        observer: 'disabledChanged',
+        value: false
+      }
+    },
+
+    get target() {
+      return this.host || this.parentNode;
+    },
+
+    configure: function() {
+      return {
+        boundDownHandler: this.onTargetDown.bind(this),
+        boundUpHandler: this.onTargetUp.bind(this)
+      };
+    },
+
+    ready: function() {
+      this.afterReady = true;
+      // TODO(cdata): We this as a workaround for polymer/polymer#1285
+      if (this.isAttached) {
+        this.attached();
+      }
+    },
+
+    attached: function() {
+      if (!this.afterReady) {
+        return;
+      }
+
+      this.target.addEventListener('mousedown', this.boundDownHandler);
+      this.target.addEventListener('mouseup', this.boundUpHandler);
+      this.target.addEventListener('keydown', this.boundDownHandler);
+      this.target.addEventListener('keyup', this.boundUpHandler);
+    },
+
+    detached: function() {
+      this.target.removeEventListener('mousedown', this.boundDownHandler);
+      this.target.removeEventListener('mouseup', this.boundUpHandler);
+      this.target.removeEventListener('keydown', this.boundDownHandler);
+      this.target.removeEventListener('keyup', this.boundUpHandler);
+    },
+
+    onTargetDown: function(event) {
+      if (event.type === 'keydown' &&
+          event.keyCode !== 32 &&
+          event.keyCode !== 13) {
+        return;
+      }
+
+      if (event.keyCode === 13) {
+        window.setTimeout(function() {
```

Comment:
Thanks for calling this out, guys. Initially I tried `this.async`. It led to what seemed to be un-obvious and erroneous behavior. Allow me to explain:

By default, `this.async` relies on a `MutationObserver` to simulate microtask asynchrony. While I did not expect this to be problematic, what I discovered was that it resulted in the `click` event dispatching _before_ the related `keyup` event. For platform-native buttons, the `click` always occurs _after_ the related `keyup`. We aren't following platform-native buttons to the letter, but we are trying to stay consistent with them where it seems reasonable.

Incidentally, `this.async` allows you to pass in an undocumented second argument: a millisecond value, similar to `setTimeout`'s second value. If the second value _happens to be greater than zero_, `this.async` will use `setTimeout` instead `MutationObserver` as its asynchrony primitive.

When `setTimeout` is used as the asynchrony primitive, the events are dispatched and handled in the correct order (caveat: in Chrome).

So, my choice was to either use `this.async` and pass in an ambiguous second argument (also the value would have to be `1`, not `0`, due to non-obvious implementation details of `this.async`), or to explicitly use `setTimeout`, which is the ultimate intention of the code.

Now, this whole explanation side-steps a question I should probably (but do not) have a good explanation for: why does asynchrony via MutationObserver cause the events to be dispatched in the wrong order? I will try to answer that question before this is merged.

And finally, if you have made it this far and you still think that `this.async` is a better option for any reason, I'm totally happy to change it.